### PR TITLE
[Friesian] Merge duplicate functions in FeatureTable and Table

### PR DESCRIPTION
**Pull request by hoshibara**
### 1. Why the change?

Fix [FeatureTable split and random_split are duplicate](https://github.com/intel-analytics/BigDL/issues/3599)
Fix [Merge cross_columns and cross_hash_encode](https://github.com/intel-analytics/BigDL/issues/3214)

### 2. User API changes

```python
FeatureTable.cross_hash_encode(columns: list[str], bins: int, cross_col_name: str, method='md5')
modified to
FeatureTable.cross_hash_encode(column_pairs: list[list[str]], bin_sizes: list[int], crossed_col_names: list[str], method='md5')
```

### 3. Summary of the change 

* remove duplicate code in random_split()
* give cross_hash_encode() the ability to process multi-pair columns
* split now an alias for random_split
* cross_columns now an alias for cross_hash_encode

### 4. How to test?
- [x] Unit test

### CODE REVIEW

File: python/friesian/src/bigdl/friesian/feature/table.py [hunk 1]
```diff
@@ -264,14 +263,11 @@ def random_split(self, weights, seed=None):
 
         :return A list of Tables
         """
-        for w in weights:
-            if w < 0.0:
-                invalidInputError(False,
-                                  "Weights must be positive. Found weight value: %s" % w)
-        seed = seed if seed is not None else random.randint(0, sys.maxsize)
         df_array = self.df.randomSplit(weights, seed)
         return [self._clone(df) for df in df_array]
```
Comment (hkvision) [inline python/friesian/src/bigdl/friesian/feature/table.py:267]: should return a list or tuple?

File: python/friesian/src/bigdl/friesian/feature/table.py [hunk 2]
```diff
@@ -1887,15 +1891,16 @@ def gen_target_code(cat_out):
...
 
             out_target_mean_dict = {
                 out_col: (target_col, target_mean_dict[target_col])
                 for target_col, out_col in zip(target_cols, out_col_list)
             }
-            return TargetCode(fold_df, cat_col, out_target_mean_dict), \
-                TargetCode(all_df, cat_col, out_target_mean_dict)
+            return TargetCode(fold_df, cat_col,
+                              out_target_mean_dict), TargetCode(all_df, cat_col,
+                                                                out_target_mean_dict)
```
Comment (hkvision) [inline python/friesian/src/bigdl/friesian/feature/table.py:1903]: revert this change?

File: python/friesian/src/bigdl/friesian/feature/table.py [hunk 3]
```diff
@@ -264,14 +263,11 @@ def random_split(self, weights, seed=None):
 
         :return A list of Tables
```
Comment (hkvision) [inline python/friesian/src/bigdl/friesian/feature/table.py:264]: Refer to split and enhance the doc a little bit?

File: python/friesian/src/bigdl/friesian/feature/table.py [hunk 4]
```diff
@@ -1004,31 +1000,66 @@ def hash_encode(self, columns, bins, method='md5'):
             hash_df = hash_df.withColumn(col_name, hash_int(pyspark_col(col_name)))
         return FeatureTable(hash_df)
 
-    def cross_hash_encode(self, columns, bins, cross_col_name=None, method='md5'):
+    def cross_hash_encode(self, column_pairs, bin_sizes, crossed_col_names=None, method='md5'):
```
Comment (hkvision) [inline python/friesian/src/bigdl/friesian/feature/table.py:1003]: is the `column_pair` name reasonable? We can do cross on more than 2 columns right?

Comment (hkvision) [reply python/friesian/src/bigdl/friesian/feature/table.py:1003]: what about `cross_columns`?

File: python/friesian/src/bigdl/friesian/feature/table.py [hunk 5]
```diff
@@ -1004,31 +1000,66 @@ def hash_encode(self, columns, bins, method='md5'):
...
         """
-        Hash encode for cross column(s).
+        Cross columns and hashed to specified bin size.
 
-        :param columns: a list of str, the categorical columns to be encoded as cross features.
+        :param columns: list[list[str]], list of categorical column pairs to be encoded
```
Comment (hkvision) [inline python/friesian/src/bigdl/friesian/feature/table.py:1007]: nested list of str, ...
remove pairs

File: python/friesian/src/bigdl/friesian/feature/table.py [hunk 6]
```diff
@@ -1004,31 +1000,66 @@ def hash_encode(self, columns, bins, method='md5'):
...
+               as cross features. i.e. [['a', 'b'], ['c', 'd']].
                For dense features, you need to cut them into discrete intervals beforehand.
-        :param bins: int, defined the number of equal-width bins in the range of column(s) values.
-        :param cross_col_name: str, the column name for output cross column. Default is None, and
-               in this case the default cross column name will be 'crossed_col1_col2'
-               for ['col1', 'col2'].
+        :param bins: list[int], defined the numbers of equal-width bins in the range
```
Comment (hkvision) [inline python/friesian/src/bigdl/friesian/feature/table.py:1010]: list of int

File: python/friesian/src/bigdl/friesian/feature/table.py [hunk 7]
```diff
@@ -1264,19 +1295,6 @@ def gen_string_idx(self, columns, freq_limit=None, order_by_freq=False,
     def _clone(self, df):
         return FeatureTable(df)
 
-    def cross_columns(self, crossed_columns, bucket_sizes):
```
Comment (hkvision) [inline python/friesian/src/bigdl/friesian/feature/table.py:1267]: will this change break the recsys demo notebook?

File: python/friesian/src/bigdl/friesian/feature/table.py [hunk 8]
```diff
@@ -1004,31 +1000,66 @@ def hash_encode(self, columns, bins, method='md5'):
...
+                crossed_col_name = 'crossed' + crossed_col_name
+            else:
+                invalidInputError(isinstance(crossed_col_names[i], str),
+                                  "each element in crossed_col_names should be string")
+                crossed_col_name = crossed_col_names[i]
+            crossed_hash_df = crossed_hash_df.withColumn(
+                crossed_col_name, concat(*column_pairs[i]))
+
+            crossed_hash_df = FeatureTable(crossed_hash_df).hash_encode(
+                [crossed_col_name], bin_sizes[i], method)
```
Comment (hkvision) [inline python/friesian/src/bigdl/friesian/feature/table.py:1059]: just crossed_col_name to make it clearer?

File: python/friesian/src/bigdl/friesian/feature/table.py [hunk 9]
```diff
@@ -1004,31 +1000,66 @@ def hash_encode(self, columns, bins, method='md5'):
...
+        :param columns: list[list[str]], list of categorical column pairs to be encoded
+               as cross features. i.e. [['a', 'b'], ['c', 'd']].
                For dense features, you need to cut them into discrete intervals beforehand.
-        :param bins: int, defined the number of equal-width bins in the range of column(s) values.
-        :param cross_col_name: str, the column name for output cross column. Default is None, and
-               in this case the default cross column name will be 'crossed_col1_col2'
-               for ['col1', 'col2'].
+        :param bins: list[int], defined the numbers of equal-width bins in the range
+               of columns values. i.e. [100, 200].
+        :param cross_col_name: list[str], the column names for output crossed columns.
```
Comment (hkvision) [inline python/friesian/src/bigdl/friesian/feature/table.py:1012]: list of str

Comment (hkvision) [reply python/friesian/src/bigdl/friesian/feature/table.py:1012]: crossed -> cross?

File: python/friesian/src/bigdl/friesian/feature/table.py (hunk 8 above)
Comment (hkvision) [reply python/friesian/src/bigdl/friesian/feature/table.py:1059]: crossed -> cross?

File: python/friesian/src/bigdl/friesian/feature/table.py (hunk 1 above)
Comment (hoshibara) [reply python/friesian/src/bigdl/friesian/feature/table.py:267]: In random_split comment: `:return A list of Tables`

File: python/friesian/src/bigdl/friesian/feature/table.py (hunk 3 above)
Comment (hoshibara) [reply python/friesian/src/bigdl/friesian/feature/table.py:264]: In pydoc, `split = random_split` will generate a line of doc like `split=random_split(self, weights, seed=None)`.
Is there still a need for additional doc in this case?

Comment (hkvision) [reply python/friesian/src/bigdl/friesian/feature/table.py:264]: I mean merging the doc of random split and split, to make the doc clearer,
For example
```
Randomly split the Table into multiple Tables with the provided weights for train, validation and test.
```

File: python/friesian/src/bigdl/friesian/feature/table.py [hunk 10]
```diff
@@ -1004,31 +1001,63 @@ def hash_encode(self, columns, bins, method='md5'):
...
+            if cross_col_names:
+                cross_col_names = [cross_col_names]
+
+        # check input params
+        invalidInputError(
+            isinstance(cross_columns, list) and all(isinstance(x, list) for x in cross_columns),
+            "cross_columns should be a nested list of column names")
+        invalidInputError(
+            isinstance(bin_sizes, list) and all(isinstance(x, int) for x in bin_sizes),
+            "bin_sizes should be a list of bin sizes")
```
Comment (hkvision) [inline python/friesian/src/bigdl/friesian/feature/table.py:1033]: a list of int

File: python/friesian/src/bigdl/friesian/feature/table.py [hunk 11]
```diff
@@ -1004,31 +1001,63 @@ def hash_encode(self, columns, bins, method='md5'):
...
+        invalidInputError(len(cross_columns) == len(bin_sizes),
+                          "cross_columns and bin_sizes should have the same length")
+        if cross_col_names is not None:
+            invalidInputError(isinstance(cross_col_names, list),
```
Comment (hkvision) [inline python/friesian/src/bigdl/friesian/feature/table.py:1037]: a list of string?

File: python/friesian/src/bigdl/friesian/feature/table.py [hunk 12]
```diff
@@ -1004,31 +1001,63 @@ def hash_encode(self, columns, bins, method='md5'):
...
+                cross_col_name = '_'.join(cross_columns[i])
+            else:
+                invalidInputError(isinstance(cross_col_names[i], str),
+                                  "each element in cross_col_names should be string")
+                cross_col_name = cross_col_names[i]
+            cross_hash_df = cross_hash_df.withColumn(
+                cross_col_name, concat(*cross_columns[i]))
+
+            cross_hash_df = FeatureTable(cross_hash_df).hash_encode(
+                [cross_col_name], bin_sizes[i], method).df
```
Comment (hkvision) [inline python/friesian/src/bigdl/friesian/feature/table.py:1057]: remove the list here?

File: python/friesian/src/bigdl/friesian/feature/table.py [hunk 13]
```diff
@@ -256,22 +255,20 @@ def filter(self, condition):
...
+        Randomly split table into multiple tables with the provided weights for train, validation
+        and test.
 
         :param weights: list of doubles as weights with which to split the table.
-            Weights will be normalized if they don't sum up to 1.0.
+               Weights will be normalized if they don't sum up to 1.0.
         :param seed: The seed for sampling.
 
-        :return A list of Tables
+        :return: A list of splitted table instances of same class.
```
Comment (hkvision) [inline python/friesian/src/bigdl/friesian/feature/table.py:265]: A list of Tables split by the provided weights.

**Review by hkvision — APPROVED** [review]
